File: protocol-generator/csharp/src/EntityCore.Protocol/Model/Execute.cs
using EntityCore.Protocol.Codec;

namespace EntityCore.Protocol.Model;

/// <summary>
/// A typed view over a <c>system/protocol/execute</c> entity (V7 §3.2). One of the
/// two wire message types. Carries a request id, target uri, operation, an entity
/// <c>params</c>, and — for authenticated requests — <c>author</c> and
/// <c>capability</c> reference hashes. The signature is a separate
/// <c>system/signature</c> entity found by target-matching in the envelope.
/// </summary>
internal sealed class Execute
{
    public Execute(Entity entity)
    {
        if (entity.Type != TypeNames.Execute)
        {
            throw new EntityProtocolException($"expected {TypeNames.Execute}, got '{entity.Type}'");
        }
        Entity = entity;
    }

    /// <summary>The underlying materialized entity (its content hash is the signature target).</summary>
    public Entity Entity { get; }

    public string RequestId => Ecf.RequireText(Entity.Data, "request_id");

    public string Uri => Ecf.RequireText(Entity.Data, "uri");

    public string Operation => Ecf.RequireText(Entity.Data, "operation");

    /// <summary>The <c>params</c> entity (materialized; §3.4).</summary>
    public Entity Params => Entity.FromDecoded(Ecf.Require(Entity.Data, "params"));

    /// <summary>Author identity hash, or null on a pre-auth connect request (§3.2, §4.2).</summary>
    public byte[]? Author => Ecf.OptBytes(Entity.Data, "author");

    /// <summary>Capability token hash, or null on a pre-auth connect request.</summary>
    public byte[]? Capability => Ecf.OptBytes(Entity.Data, "capability");

    /// <summary>The optional resource target (§3.2), or null when absent.</summary>
    public ResourceTarget? Resource
    {
        get
        {
            EcfValue? r = Ecf.Field(Entity.Data, "resource");
            return r is null ? null : ResourceTarget.FromEcf(r);
        }
    }

    /// <summary>
    /// Build a <c>system/protocol/execute</c> entity. <paramref name="paramsEntity"/>
    /// is spliced verbatim (fidelity). <paramref name="author"/> /
    /// <paramref name="capability"/> are omitted for connect-path requests (§4.2).
    /// </summary>
    public static Execute Build(
        string requestId,
        string uri,
        string operation,
        Entity paramsEntity,
        byte[]? author = null,
        byte[]? capability = null,
        ResourceTarget? resource = null,
        EcfValue? bounds = null)
    {
        EcfValue data = Ecf.Map(
            ("request_id", Ecf.Text(requestId)),
            ("uri", Ecf.Text(uri)),
            ("operation", Ecf.Text(operation)),
            ("resource", resource?.ToEcf()),
            ("params", new EcfValue.PreEncoded(paramsEntity.WireBytes)),
            ("bounds", bounds),
            ("author", author is null ? null : Ecf.Bytes(author)),
            ("capability", capability is null ? null : Ecf.Bytes(capability)));
        return new Execute(Entity.Create(TypeNames.Execute, data));
    }
}

File: protocol-generator/typescript/src/model/execute.ts
import { type EcfValue, ecfPreEncoded } from "../codec/ecf-value.js";
import { EntityProtocolError } from "../errors.js";
import * as Ecf from "./ecf.js";
import { Entity } from "./entity.js";
import { ResourceTarget } from "./resource-target.js";
import { TypeNames } from "./protocol-constants.js";

/**
 * A typed view over a `system/protocol/execute` entity (V7 §3.2). One of the two
 * wire message types. Carries a request id, target uri, operation, an entity
 * `params`, and — for authenticated requests — `author` and `capability`
 * reference hashes. The signature is a separate `system/signature` entity found
 * by target-matching in the envelope.
 */
export class Execute {
  /** The underlying materialized entity (its content hash is the signature target). */
  readonly entity: Entity;

  constructor(entity: Entity) {
    if (entity.type !== TypeNames.Execute) {
      throw new EntityProtocolError(`expected ${TypeNames.Execute}, got '${entity.type}'`);
    }
    this.entity = entity;
  }

  get requestId(): string {
    return Ecf.requireText(this.entity.data, "request_id");
  }

  get uri(): string {
    return Ecf.requireText(this.entity.data, "uri");
  }

  get operation(): string {
    return Ecf.requireText(this.entity.data, "operation");
  }

  /** The `params` entity (materialized; §3.4). */
  get params(): Entity {
    return Entity.fromDecoded(Ecf.require(this.entity.data, "params"));
  }

  /** Author identity hash, or null on a pre-auth connect request (§3.2, §4.2). */
  get author(): Uint8Array | null {
    return Ecf.optBytes(this.entity.data, "author");
  }

  /** Capability token hash, or null on a pre-auth connect request. */
  get capability(): Uint8Array | null {
    return Ecf.optBytes(this.entity.data, "capability");
  }

  /** The optional resource target (§3.2), or null when absent. */
  get resource(): ResourceTarget | null {
    const r = Ecf.field(this.entity.data, "resource");
    return r === null ? null : ResourceTarget.fromEcf(r);
  }

  /**
   * Build a `system/protocol/execute` entity. `paramsEntity` is spliced verbatim
   * (fidelity). `author` / `capability` are omitted for connect-path requests
   * (§4.2).
   */
  static build(args: {
    requestId: string;
    uri: string;
    operation: string;
    params: Entity;
    author?: Uint8Array | null;
    capability?: Uint8Array | null;
    resource?: ResourceTarget | null;
    bounds?: EcfValue | null;
  }): Execute {
    const data = Ecf.map(
      ["request_id", Ecf.text(args.requestId)],
      ["uri", Ecf.text(args.uri)],
      ["operation", Ecf.text(args.operation)],
      ["resource", args.resource ? args.resource.toEcf() : null],
      ["params", ecfPreEncoded(args.params.wireBytes)],
      ["bounds", args.bounds ?? null],
      ["author", args.author ? Ecf.bytes(args.author) : null],
      ["capability", args.capability ? Ecf.bytes(args.capability) : null],
    );
    return new Execute(Entity.create(TypeNames.Execute, data));
  }
}
